### PR TITLE
feat(versioning): Add `oci-helm` versioning

### PR DESCRIPTION
**Pull request by justusbunsi**
## Changes

Add a new `oci-helm` versioning which addresses the implicit transformation of `+` and `_` within Helm for version lookup and artifact reference.
A wrapping `oci-helm` datasource follows and is later used within Flux manager.

## Context

- #22601

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by justusbunsi** [comment]
Uh, something's off with my CLA signing. I though I already signed it. Will check...

**Comment by justusbunsi** [comment]
Right now, I am still unsure which versioning I should use as base for this, hence the missing tests at the moment. Helm claims to be compatible with SemVer 2.0, but also allows versions like `18.04` and other docker-like versions. When I compare https://regex101.com/r/vkijKf/1/ (official SemVer regex [from their docs](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string)) and the Regex that is [used within Helm](https://github.com/Masterminds/semver/blob/2f39fdc11c33c38e8b8b15b1f04334ba84e751f2/version.go#L42), I tend implement a completely new versioning based on the Helm regex. This one is used while packaging the chart as tar.gz file and defines valid versions.

Is the above a reasonable approach or am I missing something regarding SemVer within Renovate?

**Comment by justusbunsi** [comment]
> Right now, I am still unsure which versioning I should use as base for this, hence the missing tests at the moment. Helm claims to be compatible with SemVer 2.0, but also allows versions like `18.04` and other docker-like versions. When I compare https://regex101.com/r/vkijKf/1/ (official SemVer regex [from their docs](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string)) and the Regex that is [used within Helm](https://github.com/Masterminds/semver/blob/2f39fdc11c33c38e8b8b15b1f04334ba84e751f2/version.go#L42), I tend to implement a completely new versioning based on the Helm regex. This one is used while packaging the chart as tgz file and defines valid versions.
> 
> Is the above a reasonable approach or am I missing something regarding SemVer within Renovate?

I've decided to implement a versioning based on the Helm specific regex. During research I found that Helm internally depends on https://github.com/Masterminds/semver which support two types of versioning: a strict semver one and one that accept more loose values. It seems that Helm uses both versioning types for different actions:

- _helm package_ allows values like `18.04.1` (see [validation](https://github.com/helm/helm/blob/c288f0b7bed1f6174afdcdaf8779c5d37b07d7a5/pkg/action/package.go#L108)).
- _helm push_ extracts the version from packed _Chart.yaml_ and also validates with this loose versioning.
- _helm pull_ accepts loose versioning as `--version` parameter but uses strict versioning when auto-detecting the latest available version (via `helm pull oci://<registry>/<chartname>`).

Since Renovate detects OCI Chart updates via Docker Datasource, it lists all available tags via `<registry>/v2/tags/list`. This endpoint returns all versions, no matter if semver compatible or not. That's why I have to use the Helm specific regex from [Masterminds/semver -> version.go](https://github.com/Masterminds/semver/blob/2f39fdc11c33c38e8b8b15b1f04334ba84e751f2/version.go#L42). Otherwise, Renovate wouldn't be able to provide more loose versions as updates.

**Comment by justusbunsi** [comment]
I am investigating range/constraints support right now. The allowed version constraints seem similar to `composer` but have additional ones that need special converting (e.g. `~>` is allowed for `helm pull` but composer doesn't know it). Looks like I would have to implement https://github.com/Masterminds/semver inside Renovate to really support all allowed version constraints of Helm. I've added (failing) test cases for range support in the last commit.

I can do that but it would increase the complexity of this PR a lot. Before I actually start implementing range support, I'd appreciate maintainers' feedback on it:

- Add the new versioning without range support for now to keep it simple. Implement it when requested/needed?
- Implement it in this PR as well?

Any opinions @rarkins @viceice?

**Comment by rarkins** [comment]
@justusbunsi I think it's ok to support a subset of range syntax, or even just pinned versions only, so long as it's documented clearly and we have one or more issues opened for the future work

**Comment by justusbunsi** [comment]
> @justusbunsi I think it's ok to support a subset of range syntax, or even just pinned versions only, so long as it's documented clearly and we have one or more issues opened for the future work

@rarkins Thanks for your feedback. In that case, I'd prefer using the composer range logic as a basic support for the range syntax. I've seen issue references in the docs like "<abc> is an open issue for <xyz>" in the past. I'll add something similar to the documentation of this new versioning to make it clear that some syntax is not supported yet.

File: lib/modules/versioning/oci-helm/index.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+// - https://github.com/Masterminds/semver/blob/2f39fdc11c33c38e8b8b15b1f04334ba84e751f2/version.go#L42
+// - https://regex101.com/r/vkijKf/1/ (official one from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string)
+// ...into https://regex101.com/r/Visopn/1 with group names
+const versionRegex = regEx(
+  /^^v?(?<major>[0-9]\d*)(\.(?<minor>[0-9]\d*))?(\.(?<patch>[0-9]\d*))?(?:-(?<prerelease>(?:[0-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?<metadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/
+);
+
+class OciHelmVersioningApi extends GenericVersioningApi {
+  _oci2semver(value: string): string {
+    return value?.replaceAll('_', '+');
```
Comment (viceice) [inline lib/modules/versioning/oci-helm/index.ts:34]:
```suggestion
    return value.replaceAll('_', '+');
```

File: lib/modules/versioning/oci-helm/index.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  override getNewValue({ newVersion }: NewValueConfig): string | null {
+    return newVersion ? this._oci2semver(newVersion) : null;
+  }
+
+  // TODO is "versions" the fetched list? In that case, "_" has to be replaced with "+" for correct results
+  override getSatisfyingVersion(
+    versions: string[],
+    range: string
+  ): string | null {
+    const semverVersions = versions.map((v) => (v ? this._oci2semver(v) : v));
```
Comment (viceice) [inline lib/modules/versioning/oci-helm/index.ts:67]:
```suggestion
    const semverVersions = versions.map((v) => this._oci2semver(v));
```
there should never a `null` or `undefined` in that list

File: lib/modules/versioning/oci-helm/index.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    return semver.maxSatisfying(semverVersions, range, true);
+  }
+
+  // TODO is "versions" the fetched list? In that case, "_" has to be replaced with "+" for correct results
+  override minSatisfyingVersion(
+    versions: string[],
+    range: string
+  ): string | null {
+    const semverVersions = versions.map((v) => (v ? this._oci2semver(v) : v));
```
Comment (viceice) [inline lib/modules/versioning/oci-helm/index.ts:76]:
```suggestion
    const semverVersions = versions.map((v) => this._oci2semver(v));
```

File: lib/modules/versioning/oci-helm/index.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+export const id = 'oci-helm';
+export const displayName = 'OCI Helm';
+export const urls = [
+  'https://semver.org/',
+  'https://helm.sh/docs/chart_best_practices/dependencies/#versions',
+  'https://github.com/Masterminds/semver#basic-comparisons',
+  'https://helm.sh/docs/topics/registries/#oci-feature-deprecation-and-behavior-changes-with-v380',
+];
+export const supportsRanges = true;
+export const supportedRangeStrategies: RangeStrategy[] = [
```
Comment (viceice) [inline lib/modules/versioning/oci-helm/index.ts:16]: I don't see those implemented in getNewValue

Comment (justusbunsi) [reply lib/modules/versioning/oci-helm/index.ts:16]: Initially, I had the idea to re-implement https://github.com/Masterminds/semver to support every possible Helm version range variation. But as mentioned in https://github.com/renovatebot/renovate/pull/25029#issuecomment-1808608317, I decided to go with composer as the most fitting range versioning in this PR. I was short on time that's why it's not yet adjusted. Working on it right now. 🙂

Comment (justusbunsi) [reply lib/modules/versioning/oci-helm/index.ts]: Ok, I dropped the range support for now. 🙂

File: lib/modules/versioning/oci-helm/index.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+// - https://regex101.com/r/vkijKf/1/ (official one from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string)
+// ...into https://regex101.com/r/Visopn/1 with group names
+// ...prepended by range group to properly parse range values
+const versionRegex = regEx(
+  /^^(?<range>~|<|<=|>|>=|\^)?v?(?<major>[0-9]\d*)(\.(?<minor>[0-9]\d*))?(\.(?<patch>[0-9]\d*))?(?:-(?<prerelease>(?:[0-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?<metadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/,
+);
+
+class OciHelmVersioningApi extends GenericVersioningApi {
+  private _oci2semver(value: string): string {
+    return value && value.replaceAll('_', '+');
```
[the remark this answers is not in the document]
Comment (justusbunsi) [reply lib/modules/versioning/oci-helm/index.ts:27]: According to the semver regex from semver.org (https://regex101.com/r/Ly7O1x/3/) the prerelease and build metadata cannot contain underscores. If I change any of two parts on regex101 to have an underscore, the line doesn't match.

I have also just checked with `helm package --version 1.1.1-pre1+build_meta .` or `helm package --version 1.1.1-pre_1+build-meta .` which both result in `Error: Invalid Semantic Version`. So from a Helm perspective there will never be an artifact containing underscore in prerelease or build metadata.

So this should™️ be a valid replacement.

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/versioning/oci-helm/index.ts:27]: so there can only be one underscore in any version? then we can do a simple string replace
```suggestion
    return value?.replace('_', '+');
```

File: lib/modules/versioning/oci-helm/index.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+// - https://github.com/Masterminds/semver/blob/2f39fdc11c33c38e8b8b15b1f04334ba84e751f2/version.go#L42
+// - https://regex101.com/r/vkijKf/1/ (official one from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string)
+// ...into https://regex101.com/r/Visopn/1 with group names
+// ...prepended by range group to properly parse range values
+const versionRegex = regEx(
+  /^(?<range>~|<|<=|>|>=|\^)?v?(?<major>[0-9]\d*)(\.(?<minor>[0-9]\d*))?(\.(?<patch>[0-9]\d*))?(?:-(?<prerelease>(?:[0-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?<metadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/,
+);
+
+class OciHelmVersioningApi extends GenericVersioningApi {
+  private _oci2semver(value: string): string {
```
Comment (viceice) [inline lib/modules/versioning/oci-helm/index.ts:26]:
```suggestion
  private _oci2semver(value: string | null| undefined): string {
```
why do we need the conditional below 👇? do we accept null or undefined?

File: lib/modules/versioning/oci-helm/index.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    return value && value.replaceAll('_', '+');
+  }
+
+  private _semver2oci(value: string): string {
+    return value && value.replaceAll('+', '_');
```
Comment (viceice) [inline lib/modules/versioning/oci-helm/index.ts:31]:
```suggestion
  private _semver2oci(value: string | null | undefined): string | null | undefined {
    return value?.replaceAll('+', '_');
```
can the input really null or undefined?

**Comment by HonkingGoose** [comment]
Hi @justusbunsi 👋 

Do you have time to work on this PR? It's been a while since it was updated.

**Comment by justusbunsi** [comment]
Hi @HonkingGoose, I am currently not able to continue the PR due to time constraints. As soon as I have more time again I'd love to continue, if its not superseded by someone else.

To everyone reading this comment: If you have more time than me currently, and are interested in continuing the PR, feel free to do so. I am not sure when I have enough time again.

**Pull request closed without merging**